### DISCUSSION TITLE
Best way to read & rewind input body in Rack 3.0  ?

### DISCUSSION THREAD

**Discussion by andreyruby** [rack/rack#1972] [Q&A]
Question comes from Rollbar issue https://github.com/rollbar/rollbar-gem/issues/1106

As `#rewind` has gone in 3.0 there are some issues:
- Rollbar can't read body to track it (if it was already read by the app).
- It can't rewind body to allow other middlewares to read it

I propose one hack to rewind body but it looks fragile if internals change

```ruby
     body = rack_req.body #  InputWrapper
     if body.respond_to?(:rewind)
       body.rewind
     else
       body_input = body.instance_variable_get(:@input) 
       body_input.rewind if body_input.respond_to?(:rewind) # additional check if its not rewindable
    end
```

Any advises ?

**Comment by jeremyevans** [comment]
If you need rewindable input in Rack 3, use `Rack::RewindableInput::Middleware`

**Reply by andreyruby** [reply]
I think it would be great if you have method to get original input. 

For example original input can be StringIO, Tempfile, Puma::NullIO.

I would like to check if body is rewindable, then use this feature, if its not, then forget it. 

But creating additional Tempfile for each request is overkill.

**Reply by jeremyevans** [reply]
`env['rack.input'].rewind if env['rack.input'].respond_to?(:rewind)` (where `env` is the rack environment hash).

**Reply by andreyruby** [reply]
Is something wrong with my app? 
```ruby
 env['rack.input'] => #<Rack::Lint::Wrapper::InputWrapper:0x00007fcd652d6bc8 @input=#<StringIO:0x00007fcd652b1d50>>
```
`env['rack.input']`  does not respond to `#rewind` ever as it is a `Rack::Lint::Wrapper::InputWrapper`

**Reply by ioquatix** [reply]
I guess https://github.com/rack/rack/blob/main/lib/rack/lint.rb#L405 does not support rewind, or any other method that isn't specified by the spec. This might be a bug with the lint wrapper.

**Reply by ioquatix** [reply]
@jeremyevans do you think we should allow the `InputWrapper` to proxy all methods to the underlying input body or do you think this is a valid interface to enforce at this point?

**Reply by jeremyevans** [reply]
I don't think we should proxy all methods.  The purpose of `Lint` is to ensure the application follows the rack SPEC, and calling methods on the body other than those mentioned in SPEC is outside SPEC. @aglushkov's example is why we shouldn't do this.  They are using `rewind`, which is no longer required.  The code calling `rewind` is not compatible with Rack 3, and if Lint starts proxying all methods, that issue would not be caught.

**Comment by ioquatix** [comment]
Okay, so the correct way for a middleware to handle this would be:

- Read the input into a buffer.
- Replace `rack.input` with the buffered input.

Now it has both the input and the downstream middleware can also read it.

However, this has to be done carefully - i.e. if the input was expected to be streaming - like a websocket connection or a huge upload (multi-GB), etc.

**Reply by jeremyevans** [reply]
I think the recommended way in Rack 3 if you want rewindable input is to use `Rack::RewindableInput::Middleware`.  This should not be done if you want to support arbitrary streaming input in the same application.

**Comment by dblock** [comment]
Same problem in https://github.com/ruby-grape/grape/issues/2388, still discussing the best fix.

**Comment by yegor256** [comment]
Maybe someone can help with this: https://stackoverflow.com/questions/78547207/how-to-use-rackrewindableinputmiddleware-in-a-sinatra-ruby-app
